services/venice-server/src/main/java/com/linkedin/venice/listener/ReadQuotaEnforcementHandler.java

```diff
@@ -168,17 +163,32 @@ protected static double getNodeResponsibilityForQuota(
    * Initialize token buckets for all resources in the customized view repository.
    */
   public final void init() {
-    storeRepository.registerStoreDataChangedListener(this);
-    ResourceAssignment resourceAssignment = customizedViewRepository.getResourceAssignment();
-    if (resourceAssignment == null) {
-      LOGGER.error(
-          "Null resource assignment from HelixCustomizedViewOfflinePushRepository in ReadQuotaEnforcementHandler");
-    } else {
-      for (String resource: resourceAssignment.getAssignedResources()) {
-        this.onCustomizedViewChange(resourceAssignment.getPartitionAssignment(resource));
+    try {
+      storeRepository.registerStoreDataChangedListener(this);
+      ResourceAssignment resourceAssignment = customizedViewRepository.getResourceAssignment();
+      if (resourceAssignment == null) {
+        LOGGER.error(
+            "Null resource assignment from HelixCustomizedViewOfflinePushRepository in ReadQuotaEnforcementHandler");
+      } else {
+        int initResourceCount = 0;
+        for (String resource: resourceAssignment.getAssignedResources()) {
+          if (!storeVersionBuckets.containsKey(resource)) {
+            customizedViewRepository.subscribeRoutingDataChange(resource, this);
+            this.onCustomizedViewChange(resourceAssignment.getPartitionAssignment(resource));
+            initResourceCount++;
+          }
+        }
+        LOGGER.info(
+            "Current total resource count: {}, initialized {} resources during this init call",
+            resourceAssignment.getAssignedResources().size(),
+            initResourceCount);
       }
+      this.initializedVolatile = true;
+    } finally {
+      // TODO We can consider removing the periodic check and init once the customizedViewRepository can reliably return
+      // the complete resource list right after server starts
+      scheduler.schedule(this::init, RESOURCE_INIT_INTERVAL_SECONDS, TimeUnit.SECONDS);
```

Comment:
I believe this proposed approach i.e. periodical polling and updating will fix the issue. While reading the following code, I don't know if an alternative has been considered. So, if I understand, in  `onCustomizedViewDataChange`, the function `updateResourceAssignment` categorizes the new assignments into two classes: `deleted resource` vs. `updated resource ` and invokes the registered callbacks `onCustomizedViewChange` and `onRoutingDataDeleted` respectively. Here, could we create a new category, e.g. something like `added resource` and its callbacks, then `subscribeRoutingDataChange` in that callback? 

```
  protected void onCustomizedViewDataChange(RoutingTableSnapshot routingTableSnapshot) {
...
      try (AutoCloseableLock ignored = AutoCloseableLock.of(resourceAssignmentRWLock.writeLock())) {
        try (AutoCloseableLock ignore = AutoCloseableLock.of(liveInstancesMapLock)) {
          // Update the live instances as well. Helix updates live instances in this routing data
          // changed event.
          this.liveInstancesMap = Collections.unmodifiableMap(liveInstanceSnapshot);
        }
        updates = resourceAssignment.updateResourceAssignment(newResourceAssignment);
        LOGGER.info("Updated resource assignment and live instances for .");
      }
...
      for (String kafkaTopic: updates.getUpdatedResources()) {
        PartitionAssignment partitionAssignment;
        try (AutoCloseableLock ignored = AutoCloseableLock.of(resourceAssignmentRWLock.readLock())) {
          partitionAssignment = resourceAssignment.getPartitionAssignment(kafkaTopic);
        }
        listenerManager.trigger(kafkaTopic, listener -> listener.onCustomizedViewChange(partitionAssignment));
      }
      // Notify events to the listeners which listen on deleted resources.
      for (String kafkaTopic: updates.getDeletedResource()) {
        listenerManager.trigger(kafkaTopic, listener -> listener.onRoutingDataDeleted(kafkaTopic));
      }
```
I am not so sure if the alternative could work at all, just a thought, but if it could, the benefit I can see is that we might relax the reliability requirement of the very first `init` call, i.e., even it only returns a partial or an empty result, as long as it can gradually notify CV changes to server, it can be handled correctly?

Reply:
Great suggestion, let me try it and verify it in frankenwar to see if there are any nuances or edge case with this approach. Will get back to you.